### PR TITLE
KAFKA-16792: Enable consumer unit tests that fail to fetch offsets only for new consumer with poll(0)

### PR DESCRIPTION
**Pull request by FrankYang0529**
Fix following tests for CONSUMER group protocol:

* testCurrentLag
* testFetchStableOffsetThrowInPoll
* testListOffsetShouldUpdateSubscriptions

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java [hunk 1]
```diff
@@ -2471,6 +2477,8 @@ public void testCurrentLag(GroupProtocol groupProtocol) {
 
         // poll once to update with the current metadata
         consumer.poll(Duration.ofMillis(0));
+        TestUtils.waitForCondition(() -> client.requests().stream().anyMatch(request -> request.requestBuilder().apiKey().equals(ApiKeys.FIND_COORDINATOR)),
```
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java]: interesting, makes sense how you keep a single poll, and we just have to wait a bit for the background to actually generate the request. What about encapsulating this bit to make it clearer in all places it's used, to end up with something like:
`TestUtils.waitForCondition(() -> requestGenerated(ApiKeys.FIND_COORDINATOR))`

Comment (FrankYang0529) [reply clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java]: Updated it. Thanks for the suggestion.

**Review by TaiJuWu — APPROVED** [review]
LGTM

**Review by frankvicky — APPROVED** [review]
LGTM

File: clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java [hunk 2]
```diff
@@ -2523,16 +2537,32 @@ public void testListOffsetShouldUpdateSubscriptions(GroupProtocol groupProtocol)
 
         // poll once to update with the current metadata
         consumer.poll(Duration.ofMillis(0));
+        TestUtils.waitForCondition(() -> requestGenerated(client, ApiKeys.FIND_COORDINATOR),
+                "No metadata requests sent");
         client.respond(FindCoordinatorResponse.prepareResponse(Errors.NONE, groupId, metadata.fetch().nodes().get(0)));
 
         consumer.seek(tp0, 50L);
+
+        if (groupProtocol == GroupProtocol.CONSUMER) {
```
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java]: I see how you make the test pass, and that is fine and works, but I believe we could greatly simplify this test if we get the FindCoord, OffsetFetch and FetchRequests out of the picture, which seem to me are really not relevant to this test? 

This test is all about partition offsets (end offsets stored in the leader), compared to a position set manually with seek, so not related at all with committed offsets or affected by fetch.

So, this is the simplification I'm thinking about:
1. we create the consumer without groupId -> this will ensure we don't send any OffsetFetch request. We don't really need them given that we're only playing with the partition offsets
2. we pause the partition right after assign -> this ensures that we don't issue fetch requests. We don't really need them for this test, and having them makes the test different for both consumers

With those 2 small changes, I would expect we can keep the same test for both consumer, without any specifics for the new consumer, and it would still be true to its purpose, testing what it has always tested. What do you think?

**Comment by lianetm** [comment]
Hey @FrankYang0529 , I took another look, left a comment for consideration. Thanks!

File: clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java [hunk 3]
```diff
@@ -1053,12 +1053,20 @@ public void testFetchStableOffsetThrowInCommitted(GroupProtocol groupProtocol) {
...
+        TestUtils.waitForCondition(() -> {
+            try {
+                // For CONSUMER protocol, the offset request is sent in the background thread,
+                // so we need to wait until the offset request is sent.
+                consumer.poll(Duration.ZERO);
+                return false;
+            } catch (UnsupportedVersionException e) {
+                return true;
+            }
+        }, "Failed to fetch stable offset");
```
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java]: I think this msg is not quite right. If this condition fails, it means that the "consumer failed to throw UnsupportedVersionException on poll" (we actually expect that it fails to fetch the offsets)

File: clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java [hunk 4]
```diff
@@ -2481,23 +2489,31 @@ public void testCurrentLag(GroupProtocol groupProtocol) {
...
+            return hasListOffsetRequest && fetchRequest;
+        }, "No list-offset & fetch request sent");
 
         // no error for no end offset (so unknown lag)
         assertEquals(OptionalLong.empty(), consumer.currentLag(tp0));
 
         // poll once again, which should return the list-offset response
         // and hence next call would return correct lag result
-        client.respond(listOffsetsResponse(singletonMap(tp0, 90L)));
+        Optional<ClientRequest> listOffsetRequest = client.requests().stream().filter(request -> request.requestBuilder().apiKey().equals(ApiKeys.LIST_OFFSETS)).findFirst();
```
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java]: this line seems tricky to read and we need it twice (for now). Would it be clearer to have a helper along the lines of  `findRequest(client, ApiKeys.LIST_OFFSETS)`?

File: clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java [hunk 5]
```diff
@@ -2481,23 +2489,31 @@ public void testCurrentLag(GroupProtocol groupProtocol) {
...
         assertEquals(OptionalLong.of(40L), consumer.currentLag(tp0));
         // requests: fetch
-        assertEquals(1, client.inFlightRequestCount());
+        TestUtils.waitForCondition(() -> requestGenerated(client, ApiKeys.FETCH), "No fetch request sent");
 
         // one successful fetch should update the log end offset and the position
+        Optional<ClientRequest> fetchRequest = client.requests().stream().filter(request -> request.requestBuilder().apiKey().equals(ApiKeys.FETCH)).findFirst();
         final FetchInfo fetchInfo = new FetchInfo(1L, 99L, 50L, 5);
-        client.respond(fetchResponse(singletonMap(tp0, fetchInfo)));
+        assertTrue(fetchRequest.isPresent());
```
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java]: we probably don't need this assertion given that we already checked that the request was generated on the waitForCondition on ln 2510?

**Comment by lianetm** [comment]
Hey @FrankYang0529 , some other minor comments left, almost there. Thanks!

File: clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java (hunk 5 above)
Comment (FrankYang0529) [reply clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java]: Hi @lianetm, thanks for the review and suggestion. I addressed other comments. For this one, I would like to keep it. If we remove it, there may have warning message like `'Optional.get()' without 'isPresent()' check`.

File: clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java [hunk 6]
```diff
@@ -2507,32 +2523,33 @@ public void testCurrentLag(GroupProtocol groupProtocol) {
         assertEquals(OptionalLong.of(45L), consumer.currentLag(tp0));
     }
 
-    // TODO: this test triggers a bug with the CONSUMER group protocol implementation.
-    //       The bug will be investigated and fixed so this test can use both group protocols.
     @ParameterizedTest
-    @EnumSource(value = GroupProtocol.class, names = "CLASSIC")
-    public void testListOffsetShouldUpdateSubscriptions(GroupProtocol groupProtocol) {
+    @EnumSource(GroupProtocol.class)
+    public void testListOffsetShouldUpdateSubscriptions(GroupProtocol groupProtocol) throws InterruptedException {
```
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:2549]: I guess we don't need this throw Interrupted here anymore?

File: clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java [hunk 7]
```diff
@@ -2507,32 +2523,33 @@ public void testCurrentLag(GroupProtocol groupProtocol) {
...
         // correct lag result should be returned as well
         assertEquals(OptionalLong.of(40L), consumer.currentLag(tp0));
     }
 
+    private Optional<ClientRequest> findRequest(MockClient client, ApiKeys apiKey) {
+        return client.requests().stream().filter(request -> request.requestBuilder().apiKey().equals(apiKey)).findFirst();
+    }
+
+    private boolean requestGenerated(MockClient client, ApiKeys apiKey) {
+        return client.requests().stream().anyMatch(request -> request.requestBuilder().apiKey().equals(apiKey));
```
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:2578]: could we maybe reuse and simplify here to `return findRequest(client, apiKey).isPresent()`?

**Review by lianetm — COMMENTED** [review]
Thanks for the updates @FrankYang0529 ! Couple of minor comments left.

File: clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java [hunk 8]
```diff
@@ -2507,32 +2523,33 @@ public void testCurrentLag(GroupProtocol groupProtocol) {
...
 
+        client.prepareResponse(listOffsetsResponse(singletonMap(tp0, 90L)));
         assertEquals(singletonMap(tp0, 90L), consumer.endOffsets(Collections.singleton(tp0)));
         // correct lag result should be returned as well
         assertEquals(OptionalLong.of(40L), consumer.currentLag(tp0));
     }
 
+    private Optional<ClientRequest> findRequest(MockClient client, ApiKeys apiKey) {
+        return client.requests().stream().filter(request -> request.requestBuilder().apiKey().equals(apiKey)).findFirst();
+    }
```
Comment (kirktrue) [inline clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java]: It appears that the calls to this method quickly fail if there wasn't a match for the request type. How about something like this:

```suggestion
    private ClientRequest findRequest(MockClient client, ApiKeys apiKey) {
        Optional<ClientRequest> request = client.requests().stream().filter(request -> request.requestBuilder().apiKey().equals(apiKey)).findFirst();
        assertTrue(request.isPresent(), "No " + apiKey + " request was submitted to the client... or whatever");
        return request.get();  
    }
```

File: clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java [hunk 9]
```diff
@@ -2507,32 +2523,33 @@ public void testCurrentLag(GroupProtocol groupProtocol) {
...
-        consumer = newConsumer(groupProtocol, time, client, subscription, metadata, assignor, true, groupInstanceId);
-
+        consumer = newConsumer(groupProtocol, time, client, subscription, metadata, assignor, false,
+                null, groupInstanceId, false);
         consumer.assign(singleton(tp0));
-
-        // poll once to update with the current metadata
-        consumer.poll(Duration.ofMillis(0));
-        client.respond(FindCoordinatorResponse.prepareResponse(Errors.NONE, groupId, metadata.fetch().nodes().get(0)));
-
```
Comment (kirktrue) [inline clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:2542]: Out of curiosity, why do we no longer want to ensure a find coordinator request/response occurred?

**Review by kirktrue — CHANGES_REQUESTED** [review]
Thanks for the updates, @FrankYang0529!

Just a couple of questions and a minor readability change request.

Thanks!

File: clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java (hunk 6 above)
Comment (FrankYang0529) [reply clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:2549]: Thanks for the reminder. Removed it.

File: clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java (hunk 7 above)
Comment (FrankYang0529) [reply clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:2578]: Sorry, I applied @kirktrue's suggestion, so we can't use `findRequest(client, apiKey).isPresent()` here.

File: clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java (hunk 9 above)
Comment (FrankYang0529) [reply clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:2542]: Like @lianetm's suggestion https://github.com/apache/kafka/pull/16982#discussion_r1731802528. In `testListOffsetShouldUpdateSubscriptions`, we want to check `endOffsets` function. It doesn't need group coordinator.

File: clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java [hunk 10]
```diff
@@ -1053,12 +1053,20 @@ public void testFetchStableOffsetThrowInCommitted(GroupProtocol groupProtocol) {
...
-    // TODO: this test triggers a bug with the CONSUMER group protocol implementation.
-    //       The bug will be investigated and fixed so this test can use both group protocols.
     @ParameterizedTest
-    @EnumSource(value = GroupProtocol.class, names = "CLASSIC")
-    public void testFetchStableOffsetThrowInPoll(GroupProtocol groupProtocol) {
-        assertThrows(UnsupportedVersionException.class, () -> setupThrowableConsumer(groupProtocol).poll(Duration.ZERO));
+    @EnumSource(GroupProtocol.class)
+    public void testFetchStableOffsetThrowInPoll(GroupProtocol groupProtocol) throws InterruptedException {
+        setupThrowableConsumer(groupProtocol);
+        TestUtils.waitForCondition(() -> {
```
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:1075]: this change still has me thinking. This test is about a single call to `poll(ZERO)`, that is expected to throw an exception, but interesting fact is that the exception is generated when building the request [here](https://github.com/apache/kafka/blob/70dd577286de31ef20dc4f198e95f9b9e4479b47/clients/src/main/java/org/apache/kafka/common/requests/OffsetFetchRequest.java#L146) (it does not require the actual send or response). So I wonder if the async consumer should somehow ensure that when poll returns (even with low time), it has allowed for at least one run of the background thread runOnce?

File: clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java [hunk 11]
```diff
@@ -2507,32 +2521,35 @@ public void testCurrentLag(GroupProtocol groupProtocol) {
...
         consumer.assign(singleton(tp0));
-
-        // poll once to update with the current metadata
-        consumer.poll(Duration.ofMillis(0));
-        client.respond(FindCoordinatorResponse.prepareResponse(Errors.NONE, groupId, metadata.fetch().nodes().get(0)));
-
         consumer.seek(tp0, 50L);
-        client.prepareResponse(listOffsetsResponse(singletonMap(tp0, 90L)));
 
+        client.prepareResponse(listOffsetsResponse(singletonMap(tp0, 90L)));
```
Comment (TaiJuWu) [inline clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java]: Could `client.prepareResponse` be moved to `newConsumer` before?

File: clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java (hunk 10 above)
Comment (FrankYang0529) [reply clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:1075]: Hi @lianetm, the `ApplicationEventProcessor#process(PollEvent)` only resets timer for some request managers. Even if we use `ApplicationEventHandler#addAndGet`, we still can't guarantee `runOnce` is executed. We may need `ConsumerNetworkThread` to send a `BackgroundEvent` to make sure the `runOnce` is happened. However, this approach may make the process more complex. If we want to make this, we probably can create another Jira to handle this. WDYT?

Comment (lianetm) [reply clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:1075]: Agree that it's better to discuss separately. Definitely we could consider to align the behaviour of the 2 consumers a bit more regarding poll with low timeouts and the guarantees of requests sent, but there are tradeoffs to consider as @chia7712 pointed out on his [comment](https://issues.apache.org/jira/browse/KAFKA-16792?focusedCommentId=17878298&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17878298) on the Jira (totally agree on the trade-offs).

File: clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java (hunk 11 above)
Comment (lianetm) [reply clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java]: @FrankYang0529 could you address this? I guess the suggestion comes from the recent fix to some tests that were flaky for a similar situation (https://github.com/apache/kafka/pull/17056)

Comment (FrankYang0529) [reply clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java]: Hi @lianetm and @TaiJuWu, thanks for the suggestion, but I think we can keep this line just before `consumer.endOffsets`.

In `AsyncKafkaConsumer#endOffsets`, it sends `ListOffsetsEvent` and `ApplicationEventProcessor#process(ListOffsetsEvent)` calls `OffsetsRequestManager#fetchOffsets`.

In `OffsetsRequestManager#fetchOffsets`, it builds the request and `AsyncKafkaConsumer#endOffsets` uses `ApplicationEventHandler#addAndGet` to wait for the result, so I think it's safe to put it just before `consumer.endOffsets`. WDYT? Thank you.

Comment (TaiJuWu) [reply clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java]: Thanks for your explanations.

**Comment by lianetm** [comment]
Hey @FrankYang0529 , there are some related build failures on `testCurrentLag`, could you please take a look? Also please rebase to get the latest changes from trunk (we got a lot of unrelated failures too). Thanks!
https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-16982/7/tests

**Comment by FrankYang0529** [comment]
Hi @lianetm, thanks for the reminder. Rebased it to see latest result.

File: clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java [hunk 12]
```diff
@@ -2479,23 +2487,29 @@ public void testCurrentLag(GroupProtocol groupProtocol) {
         consumer.seek(tp0, 50L);
         consumer.poll(Duration.ofMillis(0));
         // requests: list-offset, fetch
-        assertEquals(2, client.inFlightRequestCount());
```
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:2499]: similar to the reasoning that led to this fix, there is a check above for:
> assertEquals(0, client.inFlightRequestCount())

I would say that may be flaky for the new consumer, right? 

We cannot ensure that no requests will be generated just because the last api we called shouldn't generate one. Given that we called poll before (and there is a background thread running) I expect there could be requests generated for fetching and fetching offsets maybe? The point of that assert is to ensure there was no ListOffsets request generated, so I would say that we should look for that exact request type, and assert there is none. Makes sense?

File: clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java (hunk 11 above)
Comment (lianetm) [reply clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java]: makes sense, but I still have a concern with this (wonder if it would make this test flaky). Let's say the moment we get to prepare this response, there is another request that got generated, wouldn't that make that the endOffsets request will never find the response? (Not sure if I'm missing details of how prepareResponse works)

Comment (lianetm) [reply clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java]: for the record, I was concerned here mostly about fetch requests (since we have a partition assigned and not paused), but just noticed we're not calling consumer.poll (so we shouldn't expect FETCH requests to be generated)

Still, worth mentioning that without this in-flight fix https://github.com/apache/kafka/pull/17035 for Fetch, we could wrongly generate fetch requests even without polling. So enabling this test like this would be flaky for the new consumer without that fix I guess.

File: clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java (hunk 12 above)
Comment (FrankYang0529) [reply clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:2499]: Hi @lianetm, thanks for catching this. I tried to add `Thread.sleep(1000)` before `assertEquals(0, client.inFlightRequestCount())`. It looks like both `ListOffsetsRequest` and `OffsetFetchRequest` will be sent after getting `FindCoordinatorResponse`. I remove the assertion now. Do you think that we should keep this assertion for CLASSIC group protocol only? Thank you.

Comment (lianetm) [reply clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:2499]: Well definitely this bit is somewhere where the 2 consumers internasl are different, even though I see them conceptually aligned:
- what's the same in both consumers: a call to consumer lag will "signal" that it needs to retrieve the log endOffsets 
- what's internally done different in both consumers: classic will only generate the request on the next poll (on the single thread it had and didn't want to block waiting for the offsets) vs async consumer, where the background thread poll will pick up the intention already expressed in the app thread and generate the request to get end offsets.

So I would say we keep the assertion (for the classic as you suggested), and it will be helpful to show this difference in the test actually. I would add an explanation for it too: Classic consumer does not send the LIST_OFFSETS right away (requires an explicit poll), different from the new async consumer, that will send the LIST_OFFSETS request in the background thread on the next background thread poll. Makes sense? 

With all these tests we're enabling, worth running them repeatedly locally to try to spot any other flakiness similar to the ones we've been catching.

Comment (FrankYang0529) [reply clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:2499]: Hi @lianetm, I added the assertion back and comments. I run 50 rounds on my laptop and there is no error. Let's check latest CI result. Thank you.

**Comment by FrankYang0529** [comment]
Hi @lianetm, I think the test result is stable now. Could you take a look when you have time? Thank you.

File: clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java (hunk 12 above)
Comment (lianetm) [reply clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:2499]: Hey @FrankYang0529, this testCurrentLag still seems flaky even after the latest changes, I filed https://issues.apache.org/jira/browse/KAFKA-17560 with what I see and where I imagine the flakiness may be, but it needs more thinking probably. Here's a suggestion to make progress: 

- we could leave `testCurrrentLag` disabled for the new consumer on this PR (just as it was before)
- we unblock the other 2 tests, that seems to pass consistently after your changes
- we address `testCurrentLag` in a separate PR, with that jira I created (including the changes you had here, just that I think it needs more)

What do you think?

Comment (FrankYang0529) [reply clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:2499]: Hi @lianetm, thanks for filing the Jira. I would like to give "eventually updated" a try in this PR. If it still can't make the test stable, we can put it in next Jira. What do you think? Thank you.

Comment (FrankYang0529) [reply clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:2499]: With latest update, I run `I=0;  while ./gradlew cleanTest :clients:test --tests KafkaConsumerTest --rerun --fail-fast; do  (( I=$I+1 )); echo "Completed run: $I"; sleep 1; done` 100 times without error.

Screenshot 2024-09-17 at 9 26 24 AM

Comment (FrankYang0529) [reply clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:2499]: Hi @lianetm, do you think I need to rebase code, so we can have more CI running result? Or, it's good to keep current state? Thank you.

Comment (FrankYang0529) [reply clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:2499]: Hi @lianetm, I rebased the code again today and latest CI result looks good. Could you help me review it when you have time? Thank you.

File: clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java [hunk 13]
```diff
@@ -2486,68 +2492,96 @@ public void testCurrentLag(GroupProtocol groupProtocol) {
...
-        client.respond(fetchResponse(singletonMap(tp0, fetchInfo)));
+        client.respondToRequest(fetchRequest, fetchResponse(singletonMap(tp0, fetchInfo)));
 
         final ConsumerRecords<String, String> records = (ConsumerRecords<String, String>) consumer.poll(Duration.ofMillis(1));
         assertEquals(5, records.count());
         assertEquals(55L, consumer.position(tp0));
 
         // correct lag result
-        assertEquals(OptionalLong.of(45L), consumer.currentLag(tp0));
+        // For AsyncKafkaConsumer, subscription sate is updated in background, so the result will eventually be updated.
```
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java]: typo state

File: clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java [hunk 14]
```diff
@@ -2486,68 +2492,96 @@ public void testCurrentLag(GroupProtocol groupProtocol) {
...
+        TestUtils.waitForCondition(() -> {
```
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java]: Is this change really needed? In this case we just did a successful fetch, so position is updated to 55 (ln 2541). We should be able to retrieve the lag of 45 (end offsets is already known to be 100). (Is not exactly the same case as above, where we needed to allow for the ListOffsets response to be processed in the background). Makes sense?

**Review by lianetm — COMMENTED** [review]
Hey @FrankYang0529, one more comment here. Almost there! Thanks!

File: clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java (hunk 14 above)
Comment (FrankYang0529) [reply clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java]: Yes, you're right, if `consumer.position` can get `45`, then the subscription state has already been updated. Remove `TestUtils.waitForCondition` here. Thanks.

**Review by lianetm — APPROVED** [review]
Thanks @FrankYang0529 ! LGTM.
